Text index: add array support for `hasAny/AllTokens`

ahmadov opened the pull request:
<!---AI changelog entry and formatting assistance: false-->
### Changelog category (leave one):

- Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Support array of string as an input to the `hasAnyTokens` or `hasAllTokens` functions.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### Details

Previously, only a string column was supported as an input to the `hasAnyTokens` and `hasAllTokens` functions. However, both `Array` and `Map` (via `mapKeys` and `mapValues`) input columns are already supported by the text index.

Since direct read from the text index is supported only by a handful functions, it was not possible to support the direct read for `Array` and `Map` columns with existing functions.

This change adds support for an array of string column as an input to both `hasAnyTokens` and `hasAllTokens` functions which led to having a direct read support for `Array` and `Map`.

NOTE: `Map` itself is not directly supported as an input, instead `mapKeys` or `mapValues` functions should be a wrapper around the `Map` column.

ahmadov committed:
Text index: add array support for `hasAny/AllTokens`

Previously, only a string column was supported as an input to the
`hasAnyTokens` and `hasAllTokens` functions. However, both `Array` and
`Map` (via `mapKeys` and `mapValues`) input columns are already
supported by the text index.
Since direct read from the text index is supported only by a handful
functions, it was not possible to support the direct read for `Array`
and `Map` columns.

This change adds support for an array of string column as an input to both
`hasAnyTokens` and `hasAnyAllTokens` functions which led to having a
direct read support for `Array` and `Map`.

NOTE: `Map` itself is not directly supported as an input, instead
`mapKeys` or `mapValues` functions should be a wrapper to `Map`.

ahmadov committed:
text index: add docs

ahmadov committed:
make text_index_array_support faster

Ergus commented on src/Functions/hasAnyAllTokens.cpp:
```diff
@@ -132,79 +145,166 @@ DataTypePtr FunctionHasAnyAllTokens<HasTokensTraits>::getReturnTypeImpl(const Co
+        {
+            if (auto it = tokens.find(std::string_view(token_start, token_len)); it != tokens.end())
+            {
+                mask |= (1ULL << it->second);
```
Minor: not needed braces here

Ergus commented on src/Functions/hasAnyAllTokens.cpp:
```diff
@@ -132,79 +145,166 @@ DataTypePtr FunctionHasAnyAllTokens<HasTokensTraits>::getReturnTypeImpl(const Co
-    if (needles.empty())
+    col_result.resize(input_rows_count);
+
+    if (tokens.empty())
```
See comment below

Ergus reviewed (commented):
Personally I would prefer this to be simpler. (without functor, less templates and callbacks).
Similarly in the tests, the tests may be as simple as possible to reduce abstractions layers to debug when there are issues.
Using a view to try the combinations is not ideal IMHO, but it is maybe just a personal opinion.

In general it looks fine to me.

ahmadov committed:
Remove duplicated code

The common code has been factored out to new `HasAnyTokensMatcher` and
`HasAllTokensMatcher` classes.

Ergus commented on src/Functions/hasAnyAllTokens.cpp:
```diff
@@ -184,67 +200,43 @@ struct HasAllTokensMatcher
             std::string_view input = input_string.getDataAt(current_offset + j).toView();
 
-            forEachTokenPadded(*token_extractor, input.data(), input.size(), HasAllTokensMatcher()(tokens, mask, expected_mask, [&] { col_result[i] = true; }));
+            forEachTokenPadded(*token_extractor, input.data(), input.size(), matcher([&] { col_result[i] = true; }));
```
Just a comment here about performance. Looking what this does, the order of these loops is extremely inefficient. Because this applies the matcher (with a callback) in the inner of the 3 loop. But remember that the most common case is that the entries in an array will be a single token. So we will be moving the heavy callback machinery like while `j < array_size; ++j`

IF our array is stored internally in contiguous memory (which probably it does), probably we should try to use the underlying `data()` to pretend that the whole array is a single longer string and reduce the calls to `forEachTokenPadded` to 1.

ahmadov replied:
If I understood correctly, you mean we can join all values in the array by a space (not physically merging but it will have the same effect) and perform a match to reduce number of calls. Unfortunately, this only make sense for the `splitByNonAlpha`, but for other tokenizers it will have a side effect.

Ergus replied:
Yes I thought that. I am a bit concerned about performance implications of these loops with a callback in the inner one...

But for now let's don't worry then.

rschu1ze commented on src/Functions/hasAnyAllTokens.cpp:
```diff
@@ -234,11 +357,11 @@ ColumnPtr FunctionHasAnyAllTokens<HasTokensTraits>::executeImpl(
     if (token_extractor == nullptr)
     {
+        /// If token_extractor == nullptr, we do a full-table scan on a column without index
+        /// By default, the default tokenizer will be used to tokenizer the input column.
```
Typo: `tokenizer` --> `tokenize`

rschu1ze reviewed (approved):
Thanks. Nice to see the refactorings in the function code overall. The functor/callbacks are hard to understand but there are currently no plans to extend hasAny/AllTokens further, so it should be okay.

ahmadov committed:
address comments

fm4v commented:
Let's backport it to 25.10

Ergus commented on src/Functions/hasAnyAllTokens.cpp:
```diff
@@ -70,16 +70,29 @@ void FunctionHasAnyAllTokens<HasTokensTraits>::setSearchTokens(const std::vector
+    if (isStringOrFixedString(type))
+        return true;
+
+    if (const auto * array_type = checkAndGetDataType<DataTypeArray>(&type); array_type)
```
A question here: Does this test also covers the Maps? They are mentioned in the documentation bellow, but this check only mentions array?

Ergus commented on src/Functions/hasAnyAllTokens.cpp:
```diff
@@ -446,6 +615,59 @@ SELECT count() FROM table WHERE hasAllTokens(msg, tokens('a()d', 'splitByString'
+    tags Array(String),
+    attributes Map(String, String),
+    INDEX idx_tags (tags) TYPE text(tokenizer = splitByNonAlpha),
+    INDEX idx_attributes_keys mapKeys(attributes) TYPE text(tokenizer = array),
```
Another question here: How do you expect that this syntax could be updated with the preprocessor? Some ideas?

Ergus reviewed (commented):
Please fix the `static DefaultTokenExtractor`

rschu1ze commented:
> Let's backport it to 25.10

@fm4v No, we shouldn't backport code for an experimental feature to an LTS release.
